src/peer_manager.rs

```diff
@@ -1141,6 +1158,48 @@ impl PeerManager {
             .collect()
     }
 
+    /// Returns the `PeerId`s and names of all the peers held in the RT, and whether they are
+    /// connected via a tunnel or not.
+    pub fn get_routing_peer_details(&self) -> Vec<(PeerId, XorName, bool)> {
+        self.routing_table
+            .iter()
+            .filter_map(|name| -> Option<(PeerId, XorName, bool)> {
+                let peer = match self.peer_map.get_by_name(name) {
+                    Some(peer) => peer,
+                    None => {
+                        error!("{:?} has {} in RT, but has no entry in peer_map for it.",
+                               self,
+                               name);
```

Comment:
At this point, I'd be unsure whether to trust the `PeerManager` (in which case we should be trying to remove this peer from RT and elsewhere) or the RT (in which case we should be adding the peer to the `peer_map`).  Short of asserting, I think we're best just logging the error so we can try and fix the underlying bug rather than fixing the mistake on the fly.  The same applies to your other two comments on this function.